workflow/lib/shared/eval_segmentation.py

```diff
@@ -88,3 +94,274 @@ def plot_cell_density_heatmap(df_cells, shape="square", plate="6W", **kwargs):
+
+        # Create paths for nuclei and cells
+        nuclei_path = f"{prefix}_nuclei.tiff"
+        cells_path = f"{prefix}_cells.tiff"
+
+        # Get aligned image path by replacing /paramsearch/tsvs/ with /
+        aligned_path = base_path.replace("/paramsearch/", "/")
+        aligned_path = (
+            aligned_path.split("__paramsearch")[0] + "__illumination_corrected.tiff"
+        )
+
+        return nuclei_path, cells_path, aligned_path
+
+    param_cols = [
+        "nuclei_diameter",
+        "cell_diameter",
+        "flow_threshold",
+        "cellprob_threshold",
+    ]
+
+    df["cell_retention"] = df["final_cells"] / df["after_edge_removal_cells"]
+    df["nuclei_retention"] = df["final_nuclei"] / df["after_edge_removal_nuclei"]
+
+    metrics = [
+        "initial_nuclei",
+        "initial_cells",
+        "final_cells",
+        "final_nuclei",
+        "cell_retention",
+        "nuclei_retention",
+    ]
+
+    grouped_stats = (
+        df.groupby(param_cols)[metrics]
+        .agg(
+            {
+                "initial_nuclei": ["mean"],
+                "initial_cells": ["mean"],
+                "final_cells": ["mean"],
+                "final_nuclei": ["mean"],
+                "cell_retention": ["mean"],
+                "nuclei_retention": ["mean"],
+            }
+        )
+        .round(2)
+    )
+
+    grouped_stats.columns = [f"{col[0]}_{col[1]}" for col in grouped_stats.columns]
+    grouped_stats["measurement_count"] = df.groupby(param_cols).size()
+
+    grouped_stats["combined_score"] = (
+        grouped_stats["cell_retention_mean"]
+        * grouped_stats["nuclei_retention_mean"]
+        * grouped_stats["final_cells_mean"]
+    )
+
+    grouped_stats = grouped_stats.sort_values("combined_score", ascending=False)
+    best_params = grouped_stats.index[0]
+    best_stats = grouped_stats.iloc[0]
+
+    # Generate summary text
+    summary_lines = []
+    summary_lines.append("=== Segmentation Parameter Optimization Summary ===")
+
+    summary_lines.append("\nOptimal Parameters:")
+    summary_lines.append(f"• Nuclei Diameter: {best_params[0]:.2f}")
+    summary_lines.append(f"• Cell Diameter: {best_params[1]:.2f}")
+    summary_lines.append(f"• Flow Threshold: {best_params[2]:.2f}")
+    summary_lines.append(f"• Cell Probability Threshold: {best_params[3]:.2f}")
+
+    summary_lines.append("\nPerformance Metrics:")
+    summary_lines.append(
+        f"• Cell Retention: {best_stats['cell_retention_mean']*100:.1f}%"
+    )
+    summary_lines.append(
+        f"• Nuclei Retention: {best_stats['nuclei_retention_mean']*100:.1f}%"
+    )
+    summary_lines.append(f"• Final Cells (avg): {best_stats['final_cells_mean']:.0f}")
+    summary_lines.append(f"• Final Nuclei (avg): {best_stats['final_nuclei_mean']:.0f}")
+    summary_lines.append(f"• Number of measurements: {best_stats['measurement_count']}")
+    summary_lines.append(f"• Combined Score: {best_stats['combined_score']:.1f}")
```

Comment:
Is there a way to do this type of text (similar stuff below) in a multi-line string? Would probably be a lot cleaner.

Reply:
Changed [here](https://github.com/cheeseman-lab/brieflow/pull/23/commits/b76ebbb6409aaa3cf569f5b5b49d1815b1b8d2f6). Good point, although still wordy.